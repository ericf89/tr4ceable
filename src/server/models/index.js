import Router from 'express-promise-router';
import userRoutes from './user/routes';
import authRoutes from './auth/routes';

const router = Router();

router.use('/auth', authRoutes);
router.use('/users', userRoutes);

export default router;

