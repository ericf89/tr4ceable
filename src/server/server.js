import express from 'express';
import bodyParser from 'body-parser';
import config from 'config';
import compression from 'compression';
import morgan from 'morgan';
import winston from 'winston';
import mongoose from 'mongoose';
import apiRoutes from './models';

mongoose.Promise = Promise;

export const mongoConnect = async () => {
  await mongoose.connect(config.get('mongoUrl'));
  return async () => {
    if (config.get('nodeEnv') === 'test') {
      await mongoose.connection.dropDatabase();
    }
    return mongoose.disconnect();
  };
};

export const server = () => {
  const app = express();

  app.set('port', config.get('port'));
  app.disable('x-powered-by');
  app.use(bodyParser.json());
  app.use(compression());

  app.use('/api', apiRoutes);
  app.get('*', (req, res) => res.sendStatus(404));
  return app;
};

export default (openDbConnection = mongoConnect, serverConfig = server) => {
  openDbConnection();
  const app = serverConfig();
  const development = config.get('nodeEnv') === 'development';
  app.locals.pretty = development;
  app.use(morgan(development ? 'dev' : 'combined'));
  app.listen(config.get('port'), () => winston.info('Listening on port %d', config.get('port')));
};

