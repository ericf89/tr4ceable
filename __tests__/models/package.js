import supertest from 'supertest';
import request from 'request-promise';
import { server, mongoConnect } from '../../src/server/server';
import User from '../../src/server/models/user/db';

jest.mock('request-promise');

const successXmlFull = `
  <?xml version="1.0" encoding="UTF-8"?>
  <TrackResponse>
      <TrackInfo ID="9405511699000600435229">
          <TrackSummary>Your item has been delivered to the mail room at 1:40 pm on April 22, 2017 in WACO, TX 76706.</TrackSummary>
          <TrackDetail>Out for Delivery, April 22, 2017, 8:03 am, WACO, TX 76706</TrackDetail>
          <TrackDetail>Sorting Complete, April 22, 2017, 7:53 am, WACO, TX 76706</TrackDetail>
          <TrackDetail>Arrived at Post Office, April 22, 2017, 5:33 am, WACO, TX 76706</TrackDetail>
          <TrackDetail>Arrived at USPS Destination Facility, April 22, 2017, 12:36 am, FORT WORTH, TX 76161</TrackDetail>
          <TrackDetail>Departed USPS Facility, April 21, 2017, 8:43 am, WEATHERFORD, TX 76086</TrackDetail>
          <TrackDetail>Arrived at USPS Facility, April 21, 2017, 8:41 am, WEATHERFORD, TX 76086</TrackDetail>
          <TrackDetail>Departed USPS Facility, April 20, 2017, 9:03 pm, FORT WORTH, TX 76161</TrackDetail>
          <TrackDetail>Arrived at USPS Facility, April 20, 2017, 8:14 am, FORT WORTH, TX 76161</TrackDetail>
          <TrackDetail>Departed USPS Facility, April 20, 2017, 3:36 am, AUSTIN, TX 78710</TrackDetail>
          <TrackDetail>Arrived at USPS Facility, April 19, 2017, 11:49 pm, AUSTIN, TX 78710</TrackDetail>
          <TrackDetail>Accepted at USPS Origin Facility, April 19, 2017, 10:34 pm, AUSTIN, TX 78701</TrackDetail>
      </TrackInfo>
  </TrackResponse>
`;

const successXmlPartial = `
  <TrackResponse>
      <TrackInfo ID="9405511699000600435229">
          <TrackSummary>Your item is out for delivery!</TrackSummary>
          <TrackDetail>Departed USPS Facility, April 20, 2017, 9:03 pm, FORT WORTH, TX 76161</TrackDetail>
          <TrackDetail>Arrived at USPS Facility, April 20, 2017, 8:14 am, FORT WORTH, TX 76161</TrackDetail>
          <TrackDetail>Departed USPS Facility, April 20, 2017, 3:36 am, AUSTIN, TX 78710</TrackDetail>
          <TrackDetail>Arrived at USPS Facility, April 19, 2017, 11:49 pm, AUSTIN, TX 78710</TrackDetail>
          <TrackDetail>Accepted at USPS Origin Facility, April 19, 2017, 10:34 pm, AUSTIN, TX 78701</TrackDetail>
      </TrackInfo>
  </TrackResponse>
`;


describe('package', () => {
  let disconnect;
  let user;
  beforeAll(async () => (disconnect = await mongoConnect()));
  beforeEach(async () => {
    await User.find().remove();
    user = await User.create({ email: 'eric@eric.com' });
  });
  afterAll(() => disconnect());

  it('should 201 add/fetch a package when posting a tracking number', async () => {
    request.mockReturnValue(Promise.resolve(successXmlFull));

    const resp = await supertest(server())
      .post(`/api/users/${user.id}/packages`)
      .send({ trackingNumber: '123456' });

    expect(request).toHaveBeenCalled();
    expect(resp.statusCode).toBe(201);
    expect(resp.body.packages && resp.body.packages.length);
    expect(resp.body.packages[0].details).toMatchSnapshot();
    expect(resp.body.packages[0].summary).toMatchSnapshot();
  });

  it('should 200 & gracefully handle a malformed response from usps', async () => {
    request.mockReturnValue(Promise.resolve('<someWeird><xmlJunk></xmlJunk></someWeird>'));

    const resp = await supertest(server())
      .post(`/api/users/${user.id}/packages`)
      .send({ trackingNumber: '123456' });

    expect(request).toHaveBeenCalled();
    expect(resp.statusCode).toBe(201);
    expect(resp.body.packages && resp.body.packages.length);
    expect(resp.body.packages[0]._id).toBeTruthy();
    expect(resp.body.packages[0].summary).toMatchSnapshot();
  });

  it('should 200 update a package when PUT to', async () => {
    request
      .mockReturnValueOnce(Promise.resolve(successXmlPartial))
      .mockReturnValueOnce(Promise.resolve(successXmlFull));
    user.packages.push({ trackingNumber: '9405511699000600435229' });
    const packageId = user.packages[0].id;
    await user.save();

    const userJSON = user.toJSON();
    expect(userJSON.packages[0].details).toEqual([]);
    expect(userJSON.packages[0].summary).toBe(undefined);

    let resp = await supertest(server())
      .put(`/api/users/${user.id}/packages/${packageId}`);

    expect(resp.statusCode).toBe(200);
    expect(resp.body.details).toMatchSnapshot();
    expect(resp.body.summary).toMatchSnapshot();

    resp = await supertest(server())
      .put(`/api/users/${user.id}/packages/${packageId}`);

    expect(resp.statusCode).toBe(200);
    expect(resp.body.details).toMatchSnapshot();
    expect(resp.body.summary).toMatchSnapshot();
  });

  it('should 404 if PUT to a nonexistent package', async () => {
    const { body, statusCode } = await supertest(server())
      .put(`/api/users/${user.id}/packages/someFakePackaroo`);

    expect(statusCode).toBe(404);
    expect(body).toMatchSnapshot();
  });

  it('should 200 when fetching a specific package by id', async () => {
    user.packages.push({ trackingNumber: '9405511699000600435229' });
    const packageId = user.packages[0].id;
    await user.save();

    const { body, statusCode } = await supertest(server())
      .get(`/api/users/${user.id}/packages/${packageId}`);

    expect(statusCode).toBe(200);
    expect(body._id).toBe(packageId);
    expect(body.summary).toMatchSnapshot();
    expect(body.details).toMatchSnapshot();
  });

  it('should 200 delete package when del to', async () => {
    user.packages.push({ trackingNumber: '9405511699000600435229' });
    const packageId = user.packages[0].id;
    await user.save();

    const resp = await supertest(server())
      .delete(`/api/users/${user.id}/packages/${packageId}`);

    expect(resp.statusCode).toBe(200);
    const updatedUser = User.findById(user.id);
    expect(updatedUser.packages).toBe(undefined);
  });

  it('should 404 when deleting a nonexistent package', async () => {
    const { body, statusCode } = await supertest(server())
      .delete(`/api/users/${user.id}/packages/someFakePackaroo`);

    expect(statusCode).toBe(404);
    expect(body).toMatchSnapshot();
  });
});
