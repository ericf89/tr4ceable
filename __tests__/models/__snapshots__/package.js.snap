// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`package should 200 & gracefully handle a malformed response from usps 1`] = `"Something went wrong.  Try again."`;

exports[`package should 200 update a package when PUT to 1`] = `
Array [
  "Departed USPS Facility, April 20, 2017, 9:03 pm, FORT WORTH, TX 76161",
  "Arrived at USPS Facility, April 20, 2017, 8:14 am, FORT WORTH, TX 76161",
  "Departed USPS Facility, April 20, 2017, 3:36 am, AUSTIN, TX 78710",
  "Arrived at USPS Facility, April 19, 2017, 11:49 pm, AUSTIN, TX 78710",
  "Accepted at USPS Origin Facility, April 19, 2017, 10:34 pm, AUSTIN, TX 78701",
]
`;

exports[`package should 200 update a package when PUT to 2`] = `"Your item is out for delivery!"`;

exports[`package should 200 update a package when PUT to 3`] = `
Array [
  "Out for Delivery, April 22, 2017, 8:03 am, WACO, TX 76706",
  "Sorting Complete, April 22, 2017, 7:53 am, WACO, TX 76706",
  "Arrived at Post Office, April 22, 2017, 5:33 am, WACO, TX 76706",
  "Arrived at USPS Destination Facility, April 22, 2017, 12:36 am, FORT WORTH, TX 76161",
  "Departed USPS Facility, April 21, 2017, 8:43 am, WEATHERFORD, TX 76086",
  "Arrived at USPS Facility, April 21, 2017, 8:41 am, WEATHERFORD, TX 76086",
  "Departed USPS Facility, April 20, 2017, 9:03 pm, FORT WORTH, TX 76161",
  "Arrived at USPS Facility, April 20, 2017, 8:14 am, FORT WORTH, TX 76161",
  "Departed USPS Facility, April 20, 2017, 3:36 am, AUSTIN, TX 78710",
  "Arrived at USPS Facility, April 19, 2017, 11:49 pm, AUSTIN, TX 78710",
  "Accepted at USPS Origin Facility, April 19, 2017, 10:34 pm, AUSTIN, TX 78701",
]
`;

exports[`package should 200 update a package when PUT to 4`] = `"Your item has been delivered to the mail room at 1:40 pm on April 22, 2017 in WACO, TX 76706."`;

exports[`package should 200 when fetching a specific package by id 1`] = `undefined`;

exports[`package should 200 when fetching a specific package by id 2`] = `Array []`;

exports[`package should 201 add/fetch a package when posting a tracking number 1`] = `
Array [
  "Out for Delivery, April 22, 2017, 8:03 am, WACO, TX 76706",
  "Sorting Complete, April 22, 2017, 7:53 am, WACO, TX 76706",
  "Arrived at Post Office, April 22, 2017, 5:33 am, WACO, TX 76706",
  "Arrived at USPS Destination Facility, April 22, 2017, 12:36 am, FORT WORTH, TX 76161",
  "Departed USPS Facility, April 21, 2017, 8:43 am, WEATHERFORD, TX 76086",
  "Arrived at USPS Facility, April 21, 2017, 8:41 am, WEATHERFORD, TX 76086",
  "Departed USPS Facility, April 20, 2017, 9:03 pm, FORT WORTH, TX 76161",
  "Arrived at USPS Facility, April 20, 2017, 8:14 am, FORT WORTH, TX 76161",
  "Departed USPS Facility, April 20, 2017, 3:36 am, AUSTIN, TX 78710",
  "Arrived at USPS Facility, April 19, 2017, 11:49 pm, AUSTIN, TX 78710",
  "Accepted at USPS Origin Facility, April 19, 2017, 10:34 pm, AUSTIN, TX 78701",
]
`;

exports[`package should 201 add/fetch a package when posting a tracking number 2`] = `"Your item has been delivered to the mail room at 1:40 pm on April 22, 2017 in WACO, TX 76706."`;

exports[`package should 404 if PUT to a nonexistent package 1`] = `Object {}`;

exports[`package should 404 when deleting a nonexistent package 1`] = `Object {}`;
